File: backend/storage.py
"""JSON-based storage for conversations."""

import json
import os
from datetime import datetime
from typing import List, Dict, Any, Optional
from pathlib import Path
from .config import DATA_DIR


def ensure_data_dir():
    """Ensure the data directory exists."""
    Path(DATA_DIR).mkdir(parents=True, exist_ok=True)


def get_conversation_path(conversation_id: str) -> str:
    """Get the file path for a conversation."""
    return os.path.join(DATA_DIR, f"{conversation_id}.json")


def create_conversation(conversation_id: str) -> Dict[str, Any]:
    """
    Create a new conversation.

    Args:
        conversation_id: Unique identifier for the conversation

    Returns:
        New conversation dict
    """
    ensure_data_dir()

    conversation = {
        "id": conversation_id,
        "created_at": datetime.utcnow().isoformat(),
        "title": "New Conversation",
        "messages": []
    }

    # Save to file
    path = get_conversation_path(conversation_id)
    with open(path, 'w') as f:
        json.dump(conversation, f, indent=2)

    return conversation


def get_conversation(conversation_id: str) -> Optional[Dict[str, Any]]:
    """
    Load a conversation from storage.

    Args:
        conversation_id: Unique identifier for the conversation

    Returns:
        Conversation dict or None if not found
    """
    path = get_conversation_path(conversation_id)

    if not os.path.exists(path):
        return None

    with open(path, 'r') as f:
        return json.load(f)


def save_conversation(conversation: Dict[str, Any]):
    """
    Save a conversation to storage.

    Args:
        conversation: Conversation dict to save
    """
    ensure_data_dir()

    path = get_conversation_path(conversation['id'])
    with open(path, 'w') as f:
        json.dump(conversation, f, indent=2)


def list_conversations(include_archived: bool = False) -> List[Dict[str, Any]]:
    """
    List all conversations (metadata only).

    Args:
        include_archived: If True, include archived conversations

    Returns:
        List of conversation metadata dicts
    """
    ensure_data_dir()

    conversations = []
    for filename in os.listdir(DATA_DIR):
        if filename.endswith('.json'):
            path = os.path.join(DATA_DIR, filename)
            with open(path, 'r') as f:
                data = json.load(f)
                # Skip archived unless requested
                if not include_archived and data.get("archived", False):
                    continue
                # Return metadata only
                conversations.append({
                    "id": data["id"],
                    "created_at": data["created_at"],
                    "title": data.get("title", "New Conversation"),
                    "message_count": len(data["messages"]),
                    "archived": data.get("archived", False)
                })

    # Sort by creation time, newest first
    conversations.sort(key=lambda x: x["created_at"], reverse=True)

    return conversations


def add_user_message(conversation_id: str, content: str):
    """
    Add a user message to a conversation.

    Args:
        conversation_id: Conversation identifier
        content: User message content
    """
    conversation = get_conversation(conversation_id)
    if conversation is None:
        raise ValueError(f"Conversation {conversation_id} not found")

    conversation["messages"].append({
        "role": "user",
        "content": content
    })

    save_conversation(conversation)


def add_assistant_message(
    conversation_id: str,
    stage1: List[Dict[str, Any]],
    stage2: List[Dict[str, Any]],
    stage3: Dict[str, Any]
):
    """
    Add an assistant message with all 3 stages to a conversation.

    Args:
        conversation_id: Conversation identifier
        stage1: List of individual model responses
        stage2: List of model rankings
        stage3: Final synthesized response
    """
    conversation = get_conversation(conversation_id)
    if conversation is None:
        raise ValueError(f"Conversation {conversation_id} not found")

    conversation["messages"].append({
        "role": "assistant",
        "stage1": stage1,
        "stage2": stage2,
        "stage3": stage3
    })

    save_conversation(conversation)


def update_conversation_title(conversation_id: str, title: str):
    """
    Update the title of a conversation.

    Args:
        conversation_id: Conversation identifier
        title: New title for the conversation
    """
    conversation = get_conversation(conversation_id)
    if conversation is None:
        raise ValueError(f"Conversation {conversation_id} not found")

    conversation["title"] = title
    save_conversation(conversation)


def archive_conversation(conversation_id: str, archived: bool = True):
    """
    Archive or unarchive a conversation.

    Args:
        conversation_id: Conversation identifier
        archived: True to archive, False to unarchive
    """
    conversation = get_conversation(conversation_id)
    if conversation is None:
        raise ValueError(f"Conversation {conversation_id} not found")

    conversation["archived"] = archived
    save_conversation(conversation)


def delete_conversation(conversation_id: str):
    """
    Permanently delete a conversation.

    Args:
        conversation_id: Conversation identifier
    """
    path = get_conversation_path(conversation_id)
    if os.path.exists(path):
        os.remove(path)


def get_overall_scores() -> Dict[str, Any]:
    """
    Calculate overall scores across all conversations.

    Aggregates rankings from all stage2 results across all conversations
    to produce a global leaderboard.

    Returns:
        Dict containing model scores and metadata
    """
    ensure_data_dir()

    # Track scores per model
    model_scores: Dict[str, Dict[str, Any]] = {}
    total_conversations = 0
    total_rankings = 0

    for filename in os.listdir(DATA_DIR):
        if not filename.endswith('.json'):
            continue

        path = os.path.join(DATA_DIR, filename)
        with open(path, 'r') as f:
            data = json.load(f)

        # Skip archived conversations
        if data.get("archived", False):
            continue

        # Process each message in the conversation
        for message in data.get("messages", []):
            if message.get("role") != "assistant":
                continue

            stage2 = message.get("stage2", [])
            if not stage2:
                continue

            total_conversations += 1

            # Get the number of models being ranked
            num_models = len(stage2)

            # Process each model's rankings
            for ranking_result in stage2:
                ranker_model = ranking_result.get("model", "unknown")
                parsed_ranking = ranking_result.get("parsed_ranking", [])

                if not parsed_ranking:
                    continue

                total_rankings += 1

                # Award points based on position (inverse ranking)
                # 1st place = num_models points, last place = 1 point
                for position, ranked_label in enumerate(parsed_ranking):
                    points = num_models - position

                    # Initialize model entry if needed
                    if ranked_label not in model_scores:
                        model_scores[ranked_label] = {
                            "model": ranked_label,
                            "total_points": 0,
                            "rankings_received": 0,
                            "first_places": 0,
                            "second_places": 0,
                            "third_places": 0,
                            "position_history": []
                        }

                    model_scores[ranked_label]["total_points"] += points
                    model_scores[ranked_label]["rankings_received"] += 1
                    model_scores[ranked_label]["position_history"].append(position + 1)

                    if position == 0:
                        model_scores[ranked_label]["first_places"] += 1
                    elif position == 1:
                        model_scores[ranked_label]["second_places"] += 1
                    elif position == 2:
                        model_scores[ranked_label]["third_places"] += 1

    # Calculate averages and sort by total points
    leaderboard = []
    for model, scores in model_scores.items():
        if scores["rankings_received"] > 0:
            scores["average_position"] = sum(scores["position_history"]) / len(scores["position_history"])
            scores["average_points"] = scores["total_points"] / scores["rankings_received"]
            del scores["position_history"]  # Remove raw data from response
            leaderboard.append(scores)

    # Sort by total points descending
    leaderboard.sort(key=lambda x: x["total_points"], reverse=True)

    return {
        "leaderboard": leaderboard,
        "total_conversations_analyzed": total_conversations,
        "total_rankings_processed": total_rankings
    }

File: lib/storage.ts
/**
 * Postgres storage for conversations using Prisma.
 */

import { prisma } from "./prisma";
import type {
  Conversation,
  ConversationMetadata,
  Message,
  Stage1Result,
  Stage2Result,
  Stage3Result,
} from "./types";

// Simple UUID generation without external dependency
function generateUUID(): string {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = Math.random() * 16 | 0;
    const v = c === 'x' ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });
}

export interface ModelScore {
  model: string;
  description?: string;
  total_points: number;
  rankings_received: number;
  first_places: number;
  second_places: number;
  third_places: number;
  average_position: number;
  average_points: number;
}

export interface OverallScores {
  leaderboard: ModelScore[];
  total_conversations_analyzed: number;
  total_rankings_processed: number;
}

/**
 * Create a new conversation.
 * @param conversationId - Optional specific ID
 * @param userId - Optional user ID for row-level security
 */
export async function createConversation(
  conversationId?: string,
  userId?: string
): Promise<Conversation> {
  const id = conversationId || generateUUID();

  const conv = await prisma.conversation.create({
    data: {
      id,
      title: "New Conversation",
      archived: false,
      user_id: userId || null,
    },
  });

  return {
    id: conv.id,
    created_at: conv.created_at.toISOString(),
    title: conv.title,
    messages: [],
  };
}

/**
 * Get a conversation with all its messages.
 * @param conversationId - Conversation ID
 * @param userId - Optional user ID for row-level security
 */
export async function getConversation(
  conversationId: string,
  userId?: string
): Promise<Conversation | null> {
  const whereClause: { id: string; user_id?: string | null } = { id: conversationId };

  // If userId provided, ensure user owns the conversation
  // If userId is undefined, we allow access (for backward compatibility with guest users)
  if (userId !== undefined) {
    whereClause.user_id = userId;
  }

  const conv = await prisma.conversation.findFirst({
    where: whereClause,
    include: {
      messages: {
        orderBy: { timestamp: "asc" },
      },
    },
  });

  if (!conv) {
    return null;
  }

  const messages: Message[] = conv.messages.map((msg) => {
    if (msg.role === "user") {
      return {
        role: "user" as const,
        content: msg.content || "",
        timestamp: msg.timestamp.toISOString(),
      };
    } else {
      return {
        role: "assistant" as const,
        stage1: msg.stage1 as unknown as Stage1Result[] | undefined,
        stage2: msg.stage2 as unknown as Stage2Result[] | undefined,
        stage3: msg.stage3 as unknown as Stage3Result | undefined,
        timestamp: msg.timestamp.toISOString(),
      };
    }
  });

  return {
    id: conv.id,
    created_at: conv.created_at.toISOString(),
    title: conv.title,
    messages,
  };
}

/**
 * List all conversations (metadata only).
 * @param includeArchived - Include archived conversations
 * @param userId - Optional user ID for row-level security
 */
export async function listConversations(
  includeArchived: boolean = false,
  userId?: string
): Promise<ConversationMetadata[]> {
  const whereClause: { archived?: boolean; user_id?: string | null } = {};

  if (!includeArchived) {
    whereClause.archived = false;
  }

  // Filter by user if userId provided
  if (userId !== undefined) {
    whereClause.user_id = userId;
  }

  const conversations = await prisma.conversation.findMany({
    where: whereClause,
    include: {
      _count: {
        select: { messages: true },
      },
    },
    orderBy: { created_at: "desc" },
  });

  return conversations.map((conv) => ({
    id: conv.id,
    created_at: conv.created_at.toISOString(),
    title: conv.title,
    message_count: conv._count.messages,
    archived: conv.archived,
  }));
}

/**
 * Add a user message to a conversation.
 */
export async function addUserMessage(
  conversationId: string,
  content: string
): Promise<void> {
  await prisma.message.create({
    data: {
      conversation_id: conversationId,
      role: "user",
      content,
    },
  });
}

/**
 * Add an assistant message with all stages to a conversation.
 */
export async function addAssistantMessage(
  conversationId: string,
  stage1: Stage1Result[],
  stage2: Stage2Result[],
  stage3: Stage3Result
): Promise<void> {
  await prisma.message.create({
    data: {
      conversation_id: conversationId,
      role: "assistant",
      stage1: stage1 as unknown as object,
      stage2: stage2 as unknown as object,
      stage3: stage3 as unknown as object,
    },
  });
}

/**
 * Update conversation title.
 */
export async function updateConversationTitle(
  conversationId: string,
  title: string
): Promise<void> {
  await prisma.conversation.update({
    where: { id: conversationId },
    data: { title },
  });
}

/**
 * Archive or unarchive a conversation.
 * @param conversationId - Conversation ID
 * @param archived - Whether to archive or unarchive
 * @param userId - Optional user ID for row-level security
 */
export async function archiveConversation(
  conversationId: string,
  archived: boolean = true,
  userId?: string
): Promise<boolean> {
  const whereClause: { id: string; user_id?: string | null } = { id: conversationId };
  if (userId !== undefined) {
    whereClause.user_id = userId;
  }

  const result = await prisma.conversation.updateMany({
    where: whereClause,
    data: { archived },
  });

  return result.count > 0;
}

/**
 * Delete a conversation permanently.
 * @param conversationId - Conversation ID
 * @param userId - Optional user ID for row-level security
 */
export async function deleteConversation(
  conversationId: string,
  userId?: string
): Promise<boolean> {
  const whereClause: { id: string; user_id?: string | null } = { id: conversationId };
  if (userId !== undefined) {
    whereClause.user_id = userId;
  }

  const result = await prisma.conversation.deleteMany({
    where: whereClause,
  });

  return result.count > 0;
}

/**
 * Calculate overall scores across all conversations.
 * Uses actual model names from Stage 1 data instead of anonymous labels.
 */
export async function getOverallScores(): Promise<OverallScores> {
  // Get all non-archived assistant messages with stage1 and stage2 data
  const allMessages = await prisma.message.findMany({
    where: {
      role: "assistant",
      conversation: { archived: false },
    },
    select: { stage1: true, stage2: true },
  });

  // Filter to only messages that have stage2 data
  const messages = allMessages.filter((m) => m.stage2 !== null);

  // Track scores per model
  const modelScores: Map<
    string,
    {
      total_points: number;
      rankings_received: number;
      first_places: number;
      second_places: number;
      third_places: number;
      position_history: number[];
    }
  > = new Map();

  let totalConversations = 0;
  let totalRankings = 0;

  for (const msg of messages) {
    const stage1Data = msg.stage1 as Stage1Result[] | null;
    const stage2Data = msg.stage2 as Stage2Result[] | null;

    if (!stage2Data || !Array.isArray(stage2Data)) continue;

    // Build label_to_model mapping from Stage 1 data
    const labelToModel: Record<string, string> = {};
    if (stage1Data && Array.isArray(stage1Data)) {
      const labels = ["A", "B", "C", "D", "E", "F", "G", "H"];
      stage1Data.forEach((result, index) => {
        if (index < labels.length) {
          labelToModel[`Response ${labels[index]}`] = result.model;
        }
      });
    }

    totalConversations++;
    const numModels = stage2Data.length;

    for (const rankingResult of stage2Data) {
      const parsedRanking = rankingResult.parsed_ranking || [];
      if (parsedRanking.length === 0) continue;

      totalRankings++;

      // Award points based on position (inverse ranking)
      for (let position = 0; position < parsedRanking.length; position++) {
        const rankedLabel = parsedRanking[position];
        // Map anonymous label to actual model name, fallback to label if not found
        const modelName = labelToModel[rankedLabel] || rankedLabel;
        const points = numModels - position;

        if (!modelScores.has(modelName)) {
          modelScores.set(modelName, {
            total_points: 0,
            rankings_received: 0,
            first_places: 0,
            second_places: 0,
            third_places: 0,
            position_history: [],
          });
        }

        const scores = modelScores.get(modelName)!;
        scores.total_points += points;
        scores.rankings_received++;
        scores.position_history.push(position + 1);

        if (position === 0) scores.first_places++;
        else if (position === 1) scores.second_places++;
        else if (position === 2) scores.third_places++;
      }
    }
  }

  // Get all model IDs that have scores
  const modelIds = Array.from(modelScores.keys());

  // Fetch descriptions from LLMModel table
  const modelDescriptions = await prisma.lLMModel.findMany({
    where: { id: { in: modelIds } },
    select: { id: true, description: true },
  });

  const descriptionMap = new Map<string, string | null>(
    modelDescriptions.map((m: { id: string; description: string | null }) => [m.id, m.description])
  );

  // Calculate averages and build leaderboard
  const leaderboard: ModelScore[] = [];
  for (const [model, scores] of modelScores) {
    if (scores.rankings_received > 0) {
      const avgPosition =
        scores.position_history.reduce((a, b) => a + b, 0) /
        scores.position_history.length;
      const avgPoints = scores.total_points / scores.rankings_received;

      leaderboard.push({
        model,
        description: descriptionMap.get(model) ?? undefined,
        total_points: scores.total_points,
        rankings_received: scores.rankings_received,
        first_places: scores.first_places,
        second_places: scores.second_places,
        third_places: scores.third_places,
        average_position: Math.round(avgPosition * 100) / 100,
        average_points: Math.round(avgPoints * 100) / 100,
      });
    }
  }

  // Sort by total points descending
  leaderboard.sort((a, b) => b.total_points - a.total_points);

  return {
    leaderboard,
    total_conversations_analyzed: totalConversations,
    total_rankings_processed: totalRankings,
  };
}
